### PR TITLE
Implement graphcast optimizer

### PR DESCRIPTION
**Pull request by ychnli**
This implements the optimization pipeline for graphcast. The main functionality is via the GraphcastForecastOptimizer, which takes a graphcast model, ground truth timeseries, loss function, optimizer, and gradient mask. 

## new features
- added gradient checkpointing in graphcast model `grid2mesh_gnn`, `mesh_gnn`, and `mesh2grid_gnn` so that gradients up to 10 days of lead time can be computed with 80GB GPU vram. 

## Known issues/bugs:
- We want to be able to only update certain parts of the input. In particular, we may want to restrict the optimization to be only over one of the input timesteps. Also, we don't want to update constants like the land sea mask. Currently, this is implemented by taking the gradient over the full input and zeroing the gradients over the subset we don't want to update. However, this is not correct, as the local gradient over the subset is not in general equal to the subset of the full gradient. We need to instead directly take the gradient over only the subset of the input we want to update. This might be hard to implement (I'm not sure), but I think an "easy" workaround is to implement a grad function that takes the gradient with respect to only one of the timesteps and use the improper zero gradient method for variables. 

- Various issues with taking gradients w.r.t. unnormalized inputs. the gradient dL/dx currently has vastly different scale for differently scaled variables e.g. geopotential vs specific humidity. I think it makes sense to compute the gradient of scaled variables instead, which we can compute by multiplying dL/dx by the standard deviation of each variable. We should use this normalized gradient as the gradient fed into the optimizer, because only then does it make sense to use the same learning rate for each variable. (see below for example)

- we should reconsider using Adam vs. just vanilla gradient descent. In particular, if the magnitude of the gradient is large compared to $\epsilon$, then Adam will produce "saturated" updates which are scalings of the sign of the gradient. (see below for example)
 
grads_and_steps_adam_debug (1)
steps_over_iter_adam_debug

### CODE REVIEW

File: src/optimize_graphcast.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+                          f"but didn't find input_best in {self.save_path}.\n"
+                          "Running inference on the last saved input instead.")
+                last_saved_input = max([int(i.split(".")[0].split("_")[-1][1:]) for i in all_input_filenames])
+                input_filenames = ["input_i000.nc", f"input_{last_saved_input:03d}.nc"]
+            else:
+                input_filenames = ["input_i000.nc", "input_best.nc"]
+        elif isinstance(which_inputs, list):
+            input_filenames = which_inputs
+        else:
+            raise ValueError("what")
```
Comment (chloeyn) [inline src/optimize_graphcast.py:274]: what

**Review by chloeyn — APPROVED** [review]
looks good to me (we already discussed changing the loss to only be computed over the later input time step)
